src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs

```diff
@@ -277,16 +222,18 @@ public class ParameterBinder
 
             Logger.DoneAttemptingToBindParameterOrProperty(parameter, modelBindingContext);
 
+            Logger.AttemptingToValidateParameterOrProperty(parameter, modelBindingContext);
+
             var modelBindingResult = modelBindingContext.Result;
 
-            if (_validatorForBackCompatOnly != null)
+            // To maintain back compat, perform top level validation only when the validator
+            // inherits from the base type.
+            var baseObjectValidator = _objectModelValidator as ObjectValidatorBase;
+            if (baseObjectValidator == null)
             {
-                // Since we don't have access to an IModelValidatorProvider, fall back
-                // on back-compatibility logic. In this scenario, top-level validation
-                // attributes will be ignored like they were historically.
```

Comment:
Restore something like this comment. Of course, correct the wording to talk about only having a legacy `IObjectModelValidator` implementation.